openeo_pg_parser_networkx/resolving_utils.py

```diff
@@ -0,0 +1,217 @@
+    '''
+    Renames process nodes so that uniqueness is guaranteed, new names
+    follow the schema root_subnode_subsubnode etc.
+    '''
+    name_remapping = [
+        (f"{process_replacement_id}_{key}", key)
+        for key, _ in process_graph[process_replacement_id].items()
+    ]
+
+    for new_key, old_key in name_remapping:
+        process_graph[process_replacement_id][new_key] = process_graph[
+            process_replacement_id
+        ].pop(old_key)
+
+
+def _adjust_parameters(process_graph, process_replacement_id, arguments):
+    '''
+    "Rewires" the from_parameter connections of filled in UDPs to be consistent with
+    the parameters available from parent nodes.
+    '''
+    if len(arguments) > 0:
+        for node_key, node in process_graph[process_replacement_id].items():
+            for arg_key, from_param in node['arguments'].items():
+                # Find from_parameter value in arguments list and replace with arguments[from_parameter value] value
+                if "from_parameter" in from_param:
+                    process_graph[process_replacement_id][node_key]['arguments'][
+                        arg_key
+                    ] = arguments[from_param['from_parameter']]
+
+
+def _adjust_references(input_graph):
+    '''
+    "Rewires" the from_node connections of filled in UDPs to be consistent with
+    the nodes available on the level of the parent nodes.
+
+    Returns the result_node of the specified level,
+    so that _remove_non_result_node can be called later.
+    '''
+    for key, node in input_graph.items():
+        if _has_children(node):
+            return_key = _adjust_references(node)
+            for _, node in input_graph.items():
+                if not _has_children(node) and 'arguments' in node:
+                    for _, argument in node['arguments'].items():
+                        stripped_key = key.split('_')[-1]
+                        if (
+                            isinstance(argument, dict)
+                            and 'from_node' in argument
+                            and (
+                                argument['from_node'] == stripped_key
+                                or argument['from_node'] == key
+                            )
+                        ):
+                            argument['from_node'] = return_key
+    return _get_result_node(input_graph)
+
+
+def _get_result_node(process_graph):
+    '''
+    Finds a result node in a process graph, if none can be found, recursively
+    enters sub-levels to find one.
+
+    If UDPs are nested singularily result nodes cannot be found on the first level and
+    thus recursion is necessary
+    '''
+    for key, node in process_graph.items():
+        if 'result' in node:
+            return key
+    for value in process_graph.values():
+        return _get_result_node(value)
+
+
+def _has_children(node):
+    '''
+    Helper function to determine if a node has children or not
```

Comment:
Actually UDPs do have process_id fields, just their implementation does not as it consists of multiple processes and is wrapped in an additional layer, this function functions to detect these additional layers, I guess a better name for it would be "is_process" as that makes its functionality clearer.

Reply:
Ahh, thanks for resolving my confusion!